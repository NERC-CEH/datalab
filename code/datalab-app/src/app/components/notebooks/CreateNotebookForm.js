import React from 'react';
import PropTypes from 'prop-types';
import { Field, reduxForm } from 'redux-form';
import { Button, Form } from 'semantic-ui-react';
import { renderTextField, renderTextArea, renderDropdownField } from '../common/form/controls';
import renderUrlTextField from '../common/form/urlTextField';
import { syncValidate, asyncValidate } from './newNotebookFormValidator';

const notebookTypes = [
  { text: 'Jupyter', value: 'jupyter' },
  { text: 'RStudio', value: 'rstudio' },
];

const CreateNotebookForm = (props) => {
  const { handleSubmit, cancel, submitting } = props;
  return (
    <Form onSubmit={ handleSubmit }>
      <Field name='displayName' label='Display Name' component={renderTextField} placeholder='Display Name' />
      <Field name='type' label='Notebook Type' component={renderDropdownField} options={notebookTypes} placeholder='Notebook Type'/>
      <Field name='name' label='URL Name' component={renderUrlTextField} placeholder='Notebook Name for URLs' />
      <Field name='description' label='Description' component={renderTextArea} placeholder='Description' />
      <Button type='submit' disabled={submitting} primary>Create</Button>
      <Button type='button' onClick={cancel}>Cancel</Button>
    </Form>
  );
};

const CreateNotebookReduxForm = reduxForm({
  form: 'createNotebook',
  validate: syncValidate,
  asyncValidate,
  asyncBlurFields: ['name'],
  destroyOnUnmount: false,
})(CreateNotebookForm);

export { CreateNotebookForm as PureCreateNotebookForm };
export default CreateNotebookReduxForm;

CreateNotebookForm.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  cancel: PropTypes.func.isRequired,
};
