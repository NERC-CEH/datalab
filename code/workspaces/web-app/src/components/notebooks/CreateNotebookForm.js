import { Field, reduxForm } from 'redux-form';
import PropTypes from 'prop-types';
import React from 'react';
import { notebookSharingOptions } from '../common/selectShareOptions';
import { formatAndParseMultiSelect, CreateFormControls, renderAdornedTextField, renderSelectField, renderTextArea, renderTextField } from '../common/form/controls';
import { getAsyncValidate, syncValidate } from './createNotebookFormValidator';
import getUrlNameStartEndText from '../../core/urlHelper';
import AssetMultiSelect from '../common/form/AssetMultiSelect';
import { useReduxFormValue } from '../../hooks/reduxFormHooks';

export const FORM_NAME = 'createNotebook';
const NAME_FIELD_NAME = 'name';
export const TYPE_FIELD_NAME = 'type';
export const VERSION_FIELD_NAME = 'version';

const commonProps = {
  component: renderTextField,
  InputLabelProps: { shrink: true },
};

const CreateNotebookForm = ({
  handleSubmit, cancel, submitting, dataStorageOptions, projectKey, typeOptions, versionOptions,
}) => {
  const typeValue = useReduxFormValue(FORM_NAME, TYPE_FIELD_NAME);
  const { startText, endText } = getUrlNameStartEndText(projectKey, window.location, typeValue);

  return (
    <form onSubmit={handleSubmit}>
      <div>
        <Field
          { ...commonProps }
          name="displayName"
          label="Display Name"
          placeholder="Display Name"
        />
      </div>
      <div>
        <Field
          { ...commonProps }
          name={TYPE_FIELD_NAME}
          label="Type"
          component={renderSelectField}
          options={typeOptions} />
      </div>
      {
        versionOptions && versionOptions.length > 0
        && <div>
          <Field
            { ...commonProps }
            name={VERSION_FIELD_NAME}
            label="Version"
            component={renderSelectField}
            options={versionOptions}
          />
        </div>
      }
      <div>
        <Field
          { ...commonProps }
          name={NAME_FIELD_NAME}
          label="URL Name"
          component={renderAdornedTextField}
          placeholder="Notebook Name for URL"
          startText={startText}
          endText={endText} />
      </div>
      <div>
        <Field
          { ...commonProps }
          name="volumeMount"
          label="Data Store to Mount"
          component={renderSelectField}
          options={dataStorageOptions}/>
      </div>
      <div>
        <Field
          { ...commonProps }
          name="description"
          label="Description"
          component={renderTextArea}
          placeholder="Description" />
      </div>
      <div>
        <Field
          { ...commonProps }
          name="shared"
          label="Sharing Status"
          component={renderSelectField}
          options={notebookSharingOptions}/>
      </div>
      <div>
        <Field
          { ...commonProps }
          name="assets"
          label="Assets"
          component={AssetMultiSelect}
          projectKey={projectKey}
          format={formatAndParseMultiSelect}
          parse={formatAndParseMultiSelect}
        />
      </div>
      <CreateFormControls onCancel={cancel} submitting={submitting} />
    </form>
  );
};

const CreateNotebookReduxForm = reduxForm({
  form: FORM_NAME,
  validate: syncValidate,
  asyncValidate: getAsyncValidate(NAME_FIELD_NAME, TYPE_FIELD_NAME),
  asyncBlurFields: [NAME_FIELD_NAME, TYPE_FIELD_NAME],
  destroyOnUnmount: false,
})(CreateNotebookForm);

export { CreateNotebookForm as PureCreateNotebookForm };
export default CreateNotebookReduxForm;

const dropDownOptionType = PropTypes.shape({
  text: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
});

const dropDownOptionArrayType = PropTypes.arrayOf(dropDownOptionType);

CreateNotebookForm.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  cancel: PropTypes.func.isRequired,
  dataStorageOptions: dropDownOptionArrayType.isRequired,
  projectKey: PropTypes.string.isRequired,
  typeOptions: dropDownOptionArrayType.isRequired,
  versionOptions: dropDownOptionArrayType.isRequired,
};
