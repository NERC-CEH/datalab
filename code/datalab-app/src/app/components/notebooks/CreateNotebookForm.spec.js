import React from 'react';
import { shallow } from 'enzyme';
import { PureCreateNotebookForm } from './CreateNotebookForm';

describe('CreateNotebookForm', () => {
  function shallowRender(props) {
    return shallow(<PureCreateNotebookForm {...props} />);
  }

  const onSubmitMock = jest.fn();

  const generateProps = () => ({
    onSubmit: onSubmitMock,
  });

  beforeEach(() => jest.resetAllMocks());

  it('creates correct snapshot for create Notebook Form', () => {
    // Arrange
    const props = generateProps();

    // Act
    const output = shallowRender(props);

    // Assert
    expect(output).toMatchSnapshot();
  });
});
