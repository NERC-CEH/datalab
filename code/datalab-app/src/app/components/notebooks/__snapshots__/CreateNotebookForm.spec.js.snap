// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateNotebookForm creates correct snapshot for create Notebook Form 1`] = `
<form>
  <div>
    <Field
      component={[Function]}
      label="Display Name"
      name="displayName"
      placeholder="Display Name"
    />
  </div>
  <div>
    <Field
      component={[Function]}
      label="Notebook Type"
      name="type"
      options={
        Array [
          Object {
            "text": "Jupyter",
            "value": "jupyter",
          },
          Object {
            "text": "Jupyterlab",
            "value": "jupyterlab",
          },
          Object {
            "text": "Rstudio",
            "value": "rstudio",
          },
          Object {
            "text": "Zeppelin",
            "value": "zeppelin",
          },
        ]
      }
    />
  </div>
  <div>
    <Field
      component={[Function]}
      endText=".datalabs.ceh.ac.uk"
      label="URL Name"
      name="name"
      placeholder="Notebook Name for URL"
      startText="http://datalab-"
    />
  </div>
  <div>
    <Field
      component={[Function]}
      label="Data Store to Mount"
      name="volumeMount"
      options={
        Array [
          Object {
            "text": "First Data Store",
            "value": "alpha",
          },
          Object {
            "text": "Second Data Store",
            "value": "beta",
          },
        ]
      }
    />
  </div>
  <div>
    <Field
      component={[Function]}
      label="Description"
      name="description"
      placeholder="Description"
    />
  </div>
  <div>
    <withStyles(Button)
      color="primary"
      raised={true}
      style={
        Object {
          "margin": 8,
        }
      }
      type="submit"
    >
      Create
    </withStyles(Button)>
    <withStyles(Button)
      onClick={[Function]}
      raised={true}
      style={
        Object {
          "margin": 8,
        }
      }
    >
      Cancel
    </withStyles(Button)>
  </div>
</form>
`;
